penzai/nn/linear_and_affine.py

```diff
@@ -771,3 +823,584 @@ class ConstantRescale(layer_base.Layer):
   def __call__(self, value: Any, **_unused_side_inputs) -> Any:
     """Scales its input by the scaling factor."""
     return jax.tree_util.tree_map(lambda x: x * self.by, value)
+
+
+def prepare_for_conv(
+    inputs: NamedArray,
+    kernel: NamedArray,
+    spatial_axis_names: Sequence[str],
+    in_axis_names: Sequence[str],
+    out_axis_names: Sequence[str],
+):
+  """Preprocess lhs and rhs for jax convolution operator"""
+
+  lhs = inputs
+  rhs = kernel
+
+  in_axis_name = "in_axis-" + "-".join(in_axis_names)
+  out_axis_name = "out_axis-" + "-".join(out_axis_names)
+
+  # merge in axes into one in channel axis for the inputs and the kernel
+  lhs = lhs.untag(*in_axis_names).flatten().tag(in_axis_name)
+  rhs = rhs.untag(*in_axis_names).flatten().tag(in_axis_name)
+
+  # merge out axes into one out channels axis for jax convolution
+  rhs = rhs.untag(*out_axis_names).flatten().tag(out_axis_name)
+
+  # untag spatial axes
+  lhs = lhs.untag(*spatial_axis_names, in_axis_name)
+  rhs = rhs.untag(*spatial_axis_names, in_axis_name, out_axis_name)
+  return lhs, rhs
+
+
+def get_named_axis_back_after_conv(
+    result: NamedArray,
+    spatial_axis_names: Sequence[str],
+    out_axis_names: Sequence[str],
+    out_axis_shape: Sequence[int],
+):
+  """Postprocess result from jax convolution operator"""
+  # Get named axes back
+  return (
+      result.tag_prefix(*spatial_axis_names)
+      .reshape(out_axis_shape)
+      .tag(*out_axis_names)
+  )
+
+
+def maybe_broadcast(value: int | Sequence[int], count: int):
+  return [value] * count if isinstance(value, int) else value
+
+
+def get_dimension_numbers(ndim):
+  return jax.lax.ConvDimensionNumbers(
+      lhs_spec=(0, ndim + 1)
+      + tuple(range(1, ndim + 1)),  # BHSpatial -> BCSpatial
+      rhs_spec=(ndim + 1, ndim) + tuple(range(ndim)),  # SpatialIO -> OISpatial
+      out_spec=(0, ndim + 1)
+      + tuple(range(1, ndim + 1)),  # BSpatialC -> BCSpatial
+  )
+
+
+@struct.pytree_dataclass
+class AbstractGeneralConv(layer_base.Layer):
```

Comment:
I'm noticing a lot of duplication between `AbstractGeneralConv`, `Conv` and `ConvTranspose`. What do you think about combining all of these into a single class and having `transpose: bool` be an attribute?

Reply:
Yeah I can see arguments either way. Other options:
- keep them as separate classes but just have a single helper function that does the whole implementation (maybe `_conv_call`), with an argument `transpose`
- implement all of the logic in AbstractGeneralConv and have an abstract class method `def _is_transposed_conv(cls) -> bool` that tells it which one to use, then just implement that method on the two subclasses